File: tests/CircleAI.Tests/MeshCapabilityRegistryTests.cs
// MeshCapabilityRegistryTests.cs
//
// (RT-12 v1) Tests for the in-memory mesh capability registry.

using System;
using System.Linq;
using System.Threading.Tasks;
using CircleAI.AetherNet;
using CircleAI.Core;
using Xunit;

namespace CircleAI.Tests;

public sealed class MeshCapabilityRegistryTests
{
    private static MeshCapabilityAdvertisement Ad(
        string peerId, string modelId, int freeKv,
        DeviceTier tier = DeviceTier.Phone,
        DateTimeOffset? at = null)
        => new(
            PeerId:              peerId,
            ModelId:             modelId,
            FreeKvTokens:        freeKv,
            Tier:                tier,
            ContextWindowTokens: 2048,
            AdvertisedAtUtc:     at ?? DateTimeOffset.UtcNow);

    [Fact]
    public async Task Upsert_ListReturnsLatest()
    {
        var reg = new InMemoryMeshCapabilityRegistry();
        await reg.UpsertAsync(Ad("p1", "Qwen3-1.7B-MNN", 1024));
        await reg.UpsertAsync(Ad("p2", "Qwen3-0.6B-MNN", 256));

        var all = reg.List();
        Assert.Equal(2, all.Count);
    }

    [Fact]
    public async Task Upsert_SamePeerReplaces()
    {
        var reg = new InMemoryMeshCapabilityRegistry();
        await reg.UpsertAsync(Ad("p1", "Qwen3-1.7B-MNN", 1024));
        await reg.UpsertAsync(Ad("p1", "Qwen3-4B-MNN", 512));

        var found = reg.Find("Qwen3-4B-MNN");
        Assert.Single(found);
        Assert.Equal(512, found[0].FreeKvTokens);
    }

    [Fact]
    public async Task Find_SortsByFreeKvDescending()
    {
        var reg = new InMemoryMeshCapabilityRegistry();
        await reg.UpsertAsync(Ad("low",    "Qwen3-1.7B-MNN", 100));
        await reg.UpsertAsync(Ad("medium", "Qwen3-1.7B-MNN", 500));
        await reg.UpsertAsync(Ad("high",   "Qwen3-1.7B-MNN", 2000));

        var found = reg.Find("Qwen3-1.7B-MNN");
        Assert.Equal(3, found.Count);
        Assert.Equal("high", found[0].PeerId);
        Assert.Equal("medium", found[1].PeerId);
        Assert.Equal("low", found[2].PeerId);
    }

    [Fact]
    public async Task Find_FiltersByMinFreeKv()
    {
        var reg = new InMemoryMeshCapabilityRegistry();
        await reg.UpsertAsync(Ad("low",  "Qwen3-1.7B-MNN", 100));
        await reg.UpsertAsync(Ad("high", "Qwen3-1.7B-MNN", 2000));

        var found = reg.Find("Qwen3-1.7B-MNN", minFreeKvTokens: 1000);
        Assert.Single(found);
        Assert.Equal("high", found[0].PeerId);
    }

    [Fact]
    public async Task Find_WrongModel_ReturnsEmpty()
    {
        var reg = new InMemoryMeshCapabilityRegistry();
        await reg.UpsertAsync(Ad("p1", "Qwen3-1.7B-MNN", 1024));
        Assert.Empty(reg.Find("Qwen3-14B-MNN"));
    }

    [Fact]
    public async Task List_StaleAfter_FiltersOldEntries()
    {
        var now = DateTimeOffset.UtcNow;
        var reg = new InMemoryMeshCapabilityRegistry
        {
            NowUtc = () => now,
        };
        await reg.UpsertAsync(Ad("fresh", "Qwen3-1.7B-MNN", 500, at: now - TimeSpan.FromSeconds(10)));
        await reg.UpsertAsync(Ad("stale", "Qwen3-1.7B-MNN", 500, at: now - TimeSpan.FromMinutes(5)));

        var live = reg.List(staleAfter: TimeSpan.FromMinutes(1));
        Assert.Single(live);
        Assert.Equal("fresh", live[0].PeerId);
    }

    [Fact]
    public async Task Remove_ReturnsTrueOnce()
    {
        var reg = new InMemoryMeshCapabilityRegistry();
        await reg.UpsertAsync(Ad("p1", "Qwen3-1.7B-MNN", 1024));
        Assert.True(await reg.RemoveAsync("p1"));
        Assert.False(await reg.RemoveAsync("p1"));
    }

    [Fact]
    public async Task NullBroadcaster_NoOps()
    {
        await NullMeshCapabilityBroadcaster.Instance.BroadcastAsync(
            Ad("p1", "Qwen3-1.7B-MNN", 1024));
        // No throw; nothing to assert.
        Assert.True(true);
    }
}

File: swift/Tests/CircleAITests/MeshCapabilityRegistryTests.swift
// MeshCapabilityRegistryTests.swift
//
// Validates the CircleAI.AetherNet mesh-capability port
// (MeshCapabilityRegistry.swift): upsert-replaces-per-peer, idempotent remove,
// stale filtering, Find (model match case-insensitive + min-KV filter + spare-
// budget-descending order), the Null broadcaster no-op, and the loopback
// broadcaster's real registry effect.

import XCTest
import Foundation
@testable import CircleAI

final class MeshCapabilityRegistryTests: XCTestCase {

    private func ad(
        _ peer: String, model: String = "Qwen3-1.7B-MNN", kv: Int = 2048,
        tier: DeviceTier = .phone, ctx: Int = 4096, at: Date, latency: Int? = nil
    ) -> MeshCapabilityAdvertisement {
        MeshCapabilityAdvertisement(
            peerId: peer, modelId: model, freeKvTokens: kv, tier: tier,
            contextWindowTokens: ctx, advertisedAtUtc: at, latencyHintMs: latency)
    }

    // ── Upsert / replace / list ───────────────────────────────────────────────

    func testUpsertReplacesPerPeer() async throws {
        let reg = InMemoryMeshCapabilityRegistry()
        let now = Date()
        try await reg.upsert(ad("p1", kv: 1000, at: now))
        try await reg.upsert(ad("p1", kv: 4000, at: now)) // replaces, does not append
        let all = reg.list()
        XCTAssertEqual(all.count, 1)
        XCTAssertEqual(all.first?.freeKvTokens, 4000)
    }

    func testUpsertRejectsBlankPeerId() async {
        let reg = InMemoryMeshCapabilityRegistry()
        do {
            try await reg.upsert(ad("   ", at: Date()))
            XCTFail("expected invalidPeerId")
        } catch let e as MeshCapabilityError {
            XCTAssertEqual(e, .invalidPeerId)
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    func testListEmptyRegistryIsEmpty() {
        let reg = InMemoryMeshCapabilityRegistry()
        XCTAssertTrue(reg.list().isEmpty)
        XCTAssertTrue(reg.list(staleAfter: 60).isEmpty)
    }

    // ── Remove ────────────────────────────────────────────────────────────────

    func testRemoveIsIdempotent() async throws {
        let reg = InMemoryMeshCapabilityRegistry()
        try await reg.upsert(ad("p1", at: Date()))
        let first = try await reg.remove(peerId: "p1")
        let second = try await reg.remove(peerId: "p1")
        XCTAssertTrue(first)
        XCTAssertFalse(second)
        XCTAssertTrue(reg.list().isEmpty)
    }

    func testRemoveRejectsBlankPeerId() async {
        let reg = InMemoryMeshCapabilityRegistry()
        do {
            _ = try await reg.remove(peerId: "")
            XCTFail("expected invalidPeerId")
        } catch let e as MeshCapabilityError {
            XCTAssertEqual(e, .invalidPeerId)
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    // ── Stale filtering (fixed clock) ─────────────────────────────────────────

    func testListStaleFilterUsesInjectedClock() async throws {
        let fixedNow = Date(timeIntervalSince1970: 10_000)
        let reg = InMemoryMeshCapabilityRegistry(nowUtc: { fixedNow })
        // Fresh: 30s ago. Stale: 120s ago.
        try await reg.upsert(ad("fresh", at: fixedNow.addingTimeInterval(-30)))
        try await reg.upsert(ad("stale", at: fixedNow.addingTimeInterval(-120)))

        let recent = reg.list(staleAfter: 60) // keep entries newer than 60s
        XCTAssertEqual(recent.count, 1)
        XCTAssertEqual(recent.first?.peerId, "fresh")

        // nil staleAfter returns everything.
        XCTAssertEqual(reg.list(staleAfter: nil).count, 2)
    }

    // ── Find ──────────────────────────────────────────────────────────────────

    func testFindMatchesModelCaseInsensitiveAndSortsBySpareBudgetDesc() async throws {
        let now = Date()
        let reg = InMemoryMeshCapabilityRegistry(nowUtc: { now })
        try await reg.upsert(ad("low",  model: "Qwen3-1.7B-MNN", kv: 512,  at: now))
        try await reg.upsert(ad("high", model: "qwen3-1.7b-mnn", kv: 8192, at: now)) // different case
        try await reg.upsert(ad("mid",  model: "Qwen3-1.7B-MNN", kv: 2048, at: now))
        try await reg.upsert(ad("other", model: "Llama-3B", kv: 9999, at: now))       // different model

        let hits = reg.find(modelId: "Qwen3-1.7B-MNN")
        XCTAssertEqual(hits.map { $0.peerId }, ["high", "mid", "low"]) // spare budget desc
        XCTAssertFalse(hits.contains { $0.peerId == "other" })
    }

    func testFindHonoursMinFreeKvTokens() async throws {
        let now = Date()
        let reg = InMemoryMeshCapabilityRegistry(nowUtc: { now })
        try await reg.upsert(ad("a", kv: 100, at: now))
        try await reg.upsert(ad("b", kv: 3000, at: now))
        let hits = reg.find(modelId: "Qwen3-1.7B-MNN", minFreeKvTokens: 1000)
        XCTAssertEqual(hits.map { $0.peerId }, ["b"])
    }

    func testFindHonoursStaleFilter() async throws {
        let now = Date(timeIntervalSince1970: 50_000)
        let reg = InMemoryMeshCapabilityRegistry(nowUtc: { now })
        try await reg.upsert(ad("fresh", kv: 100, at: now.addingTimeInterval(-10)))
        try await reg.upsert(ad("stale", kv: 9999, at: now.addingTimeInterval(-600)))
        let hits = reg.find(modelId: "Qwen3-1.7B-MNN", minFreeKvTokens: 0, staleAfter: 60)
        XCTAssertEqual(hits.map { $0.peerId }, ["fresh"])
    }

    func testFindBlankModelReturnsEmpty() async throws {
        let reg = InMemoryMeshCapabilityRegistry()
        try await reg.upsert(ad("a", at: Date()))
        XCTAssertTrue(reg.find(modelId: "  ").isEmpty)
    }

    // ── Codable round-trip ────────────────────────────────────────────────────

    func testAdvertisementCodableRoundTrip() throws {
        let a = ad("p", kv: 1234, tier: .workstation, ctx: 131_072,
                   at: Date(timeIntervalSince1970: 1), latency: 42)
        let back = try JSONDecoder().decode(
            MeshCapabilityAdvertisement.self, from: JSONEncoder().encode(a))
        XCTAssertEqual(back, a)
    }

    // ── Broadcasters ──────────────────────────────────────────────────────────

    func testNullBroadcasterIsNoOp() async throws {
        // Must not throw and must have no observable effect.
        try await NullMeshCapabilityBroadcaster.shared.broadcast(ad("p", at: Date()))
    }

    func testLoopbackBroadcasterUpsertsIntoRegistry() async throws {
        let now = Date()
        let reg = InMemoryMeshCapabilityRegistry(nowUtc: { now })
        let bc = LoopbackMeshCapabilityBroadcaster(registry: reg)
        try await bc.broadcast(ad("self", model: "M", kv: 2048, at: now))
        // Broadcasting our advert makes it discoverable via find.
        let hits = reg.find(modelId: "M")
        XCTAssertEqual(hits.map { $0.peerId }, ["self"])
    }
}
